File: Array/austenYang/Leetcode/JavaKotlin/src/L3/Solution.java
package L3;

import java.util.HashSet;

public class Solution {
    /**
     * 滑动窗口
     *
     * @param s
     * @return
     */
    public static int lengthOfLongestSubstring(String s) {
        char[] chars = s.toCharArray();
        int n = chars.length;
        HashSet<Character> occ = new HashSet<>();
        /**
         * 在滑动窗口中
         * i 表示左指针
         * rk 表示有指正 初始值为 -1 表示 它在 子字符串的左边界的左侧，还没有开始移动
         */
        int rk = -1, ans = 0;
        for (int i = 0; i < n; i++) {
            if (i != 0) {
                occ.remove(chars[i - 1]);
            }
            while (rk + 1 < n && !occ.contains(chars[rk + 1])) {
                occ.add(chars[rk + 1]);
                rk++;
            }
            ans = Math.max(ans, rk - i + 1);
        }
        return ans;
    }

    public static void main(String[] args) {
        System.out.println(lengthOfLongestSubstring("pwwkew"));
    }
}

File: Array/austenYang/Leetcode/JavaKotlin/src/L3/Solution.kt
package L3


/**
 * 窗口滑动
 */
fun lengthOfLongestSubstring(s: String): Int {
    val hashset = hashSetOf<Char>()
    var rk = -1
    var ans = 0
    val n = s.length
    s.forEachIndexed { index, c ->
        if (index != 0) {
            // 当窗口左边界移动时，那么就要依次地把窗口最左边的元素移除
            hashset.remove(s[index - 1])
        }
        while (rk + 1 < n && !hashset.contains(c)) {
            hashset.add(c)
            rk++
        }
        ans = ans.coerceAtLeast(rk - index + 1)
    }
    return ans
}

fun main() {

}